Starlight 0.39 blog post (#2381)
withastro/astro.build: src/content/blog/starlight-039.mdx

---
title: Starlight 0.39

> description: Customize icons, attributes, and page slugs with the latest Starlight release.

Reviewer: Probably an outdated description. Suggesting something but probably not that good:

```suggestion
description: The latest Starlight release brings more flexible autogenerated sidebars, improved styling, and stronger multilingual support.
```

Author: Ah, great catch, copy/pasted from the previous blog post. Fixed in a8b3ad0484ede871a496e37306e0900e9cbdaefa

> description: Get more flexible autogenerated sidebars, improved styling, and stronger multilingual docs support with the latest Starlight release.

Reviewer 2: Small wording rephrasing:

```suggestion
description: Get more flexible autogenerated sidebars, improved styling, and stronger support for multilingual docs with the latest Starlight release.
```

publishDate: 2026-05-07
authors:
  - chris
coverImage: '/src/content/blog/_images/starlight-039/cover-starlight-039.webp'
socialImage: '/src/content/blog/_images/starlight-039/og-starlight-039.webp'
related:
  - starlight-035
  - starlight-034
---

Starlight is Astro’s official package for building fast, accessible documentation websites, with all the features you need built in.

Here’s what’s new in this release:

- [**More flexible autogenerated sidebar links**](#flexible-sidebar-autogeneration)
- [**Styling improvements**](#styling-improvements)
- [**Features for multilingual sites**](#better-multilingual-sites)

To upgrade an existing Starlight site, use the automated `@astrojs/upgrade` CLI tool. This will update Starlight, Astro, and any other integrations you are using:

```sh
npx @astrojs/upgrade
```

See the [Starlight CHANGELOG](https://github.com/withastro/starlight/blob/main/packages/starlight/CHANGELOG.md) for full details of breaking changes.

## Flexible sidebar autogeneration

Starlight has included support for generating sidebar navigation links from directories of content since day one.
This release brings new flexibility to how this works.

> Previously, autogeneration always created a labelled group in the sidebar for the specified directory.

Reviewer 2: I think "labelled" is British English (little nitpicky):

```suggestion
Previously, autogeneration always created a labeled group in the sidebar for the specified directory.
```

With v0.39, Starlight’s autogenerated links can be slotted alongside any other link types.

If you have existing autogenerated link groups in your sidebar, update them by wrapping the config in an `items` array:

```diff lang="js" title="astro.config.mjs"
starlight({
	sidebar: [
		{
			label: 'My group',
-			autogenerate: { directory: 'some-dir' },
+			items: [{ autogenerate: { directory: 'some-dir' } }],
		},
	],
	// ...
}),
```

This change makes it easy to combine autogeneration with different kinds of links.
For example, you can now add one-off custom links alongside autogenerated ones.
The following example groups a local page, all pages in the `features` directory, and an external link:

```js {6-11}
starlight({
	sidebar: [
		{
			label: 'Features',
			items: [
				// A page not in the features directory
				'features-overview',
				// All pages in the features directory
				{ autogenerate: { directory: 'features' } },
				// An external link
				{ label: 'Support', link: 'https://support.example.com' },
			],
		},
	],
	// ...
}),
```

## Styling improvements

> Modern CSS makes everyone’s lives easier by the day and with this release we’re putting it to good use.

Reviewer 3: ```suggestion
Modern CSS makes everyone’s lives easier by the day, and with this release we’re putting it to good use.
```

> - We’ve improved spacing of nested lists in your content thanks to the magic of `:has()`. (Check out [the full selector](https://github.com/withastro/starlight/blob/ec70630cbbe43bb5b026ddabfe7224759c798a4b/packages/starlight/style/markdown.css#L34-L60) — it’s a doozy.)

Reviewer 2: Do you wanna add an article here?

```suggestion
- We’ve improved the spacing of nested lists in your content thanks to the magic of `:has()`. (Check out [the full selector](https://github.com/withastro/starlight/blob/ec70630cbbe43bb5b026ddabfe7224759c798a4b/packages/starlight/style/markdown.css#L34-L60) — it’s a doozy.)
```

Author: Not in this case — an article wouldn’t be wrong here, but it’s just a stylistic choice.

> - The Starlight sidebar is more stable when users expand or collapse a group causing overflow. We banished layout shift with a simple `scrollbar-gutter: stable`.

Reviewer 3: ```suggestion
- The Starlight sidebar is more stable when users expand or collapse a group, causing overflow. We banished layout shift with a simple `scrollbar-gutter: stable`.
```

## Better multilingual sites

> Starlight’s internationalization story has been key from the outset and we’re always looking to push it forward.

Reviewer 3: ```suggestion
Starlight’s internationalization story has been key from the outset, and we’re always looking to push it forward.
```

This release brings:

- Improved multilingual SEO with `x-default` meta links to pages in the site’s default locale. These alternate links are [used by search engines](https://developers.google.com/search/docs/specialty/international/localized-versions#xdefault) as a signal of which language to fall back to if no other is available.

> - Better text layout in Chinese and Japanese documentation thanks to the new `text-autospace` CSS property. This [newly baseline CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/text-autospace) makes text more legible by automatically applying spacing between CJK and non-CJK characters.

Reviewer 2: I think Baseline should be uppercase first letter since it's a web standard (own name):

```suggestion
- Better text layout in Chinese and Japanese documentation thanks to the new `text-autospace` CSS property. This [newly Baseline CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/text-autospace) makes text more legible by automatically applying spacing between CJK and non-CJK characters.
```

Author: Yeah, you’re probably right. I don’t love it for an adjective though 😅 

Maybe we can just go with this:

```suggestion
- Better text layout in Chinese and Japanese documentation thanks to the `text-autospace` CSS property. This [newer CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/text-autospace) makes text more legible by automatically applying spacing between CJK and non-CJK characters.
```

- We’ve updated `i18next`, the underlying library powering Starlight’s localization APIs, to the latest major release.

## Community

import Mention from '~/components/Mention.astro';

The Astro core team is:

<Mention name="alex" />, <Mention name="armand" />, <Mention name="chris" />,
<Mention name="ema" />,
<Mention name="erika" />, <Mention name="florian" />, <Mention name="fred" />,
<Mention name="hideoo" />,
<Mention name="luiz" />, <Mention name="matt" />, <Mention name="matthew" />,
<Mention name="reuben" />,
<Mention name="sarah" />, and <Mention name="yan" />.

> Thanks to everyone who contributed to our recent releases with PRs and reviews, including [alomtvassam-ui](https://github.com/alomtvassam-ui), [Andreas Deininger](https://github.com/deining), [Andriy Massimilla](https://github.com/andriygm), [Bugo](https://github.com/dragomano), [Elian](https://github.com/ElianCodes), [Felix Schneider](https://github.com/trueberryless), [gesslar](https://github.com/gesslar), [Ian Trewella](https://github.com/itrew), [John Goodliff](https://github.com/twocaretcat), [Joost de Valk](https://github.com/jdevalk), [Josef Ježek](https://github.com/JosefJezek), [Junseong Park](https://github.com/jsparkdev), [Kevin Z](https://github.com/kevinzunigacuellar), [Linus Grolmes](https://github.com/madLinux7), [Luke Hsiao](https://github.com/lukehsiao), [Maxime Mangel](https://github.com/MangelMaxime), [MilesChou](https://github.com/MilesChou), [Patrick Schratz](https://github.com/pat-s), [Paweł Cisło](https://github.com/pyxelr), [Roman](https://github.com/gameroman), [Sakku](https://github.com/Yusaku01), [Shinya Fujino](https://github.com/morinokami), [Swapnil Jain](https://github.com/incognitosj), [Tatsunori Uchino](https://github.com/tats-u), [vincent](https://github.com/vincent067), [Vojtěch Křižan](https://github.com/voytxt), and [Woody Chang](https://github.com/kazettique)

Reviewer: ```suggestion
Thanks to everyone who contributed to our recent releases with PRs and reviews, including [alomtvassam-ui](https://github.com/alomtvassam-ui), [Andreas Deininger](https://github.com/deining), [Andriy Massimilla](https://github.com/andriygm), [Bugo](https://github.com/dragomano), [Elian](https://github.com/ElianCodes), [Felix Schneider](https://github.com/trueberryless), [gesslar](https://github.com/gesslar), [Ian Trewella](https://github.com/itrew), [John Goodliff](https://github.com/twocaretcat), [Joost de Valk](https://github.com/jdevalk), [Josef Ježek](https://github.com/JosefJezek), [Junseong Park](https://github.com/jsparkdev), [Kevin Z](https://github.com/kevinzunigacuellar), [Linus Grolmes](https://github.com/madLinux7), [Luke Hsiao](https://github.com/lukehsiao), [Maxime Mangel](https://github.com/MangelMaxime), [MilesChou](https://github.com/MilesChou), [Patrick Schratz](https://github.com/pat-s), [Paweł Cisło](https://github.com/pyxelr), [Roman](https://github.com/gameroman), [Sakku](https://github.com/Yusaku01), [Shinya Fujino](https://github.com/morinokami), [Swapnil Jain](https://github.com/incognitosj), [Tatsunori Uchino](https://github.com/tats-u), [vincent](https://github.com/vincent067), [Vojtěch Křižan](https://github.com/voytxt), and [Woody Chang](https://github.com/kazettique).
```

Should this sentence include a final dot?
